pmd-apex/src/main/java/net/sourceforge/pmd/lang/apex/rule/bestpractices/QueueableShouldAttachFinalizerRule.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * BSD-style license; for more info see http://pmd.sourceforge.net/license.html
+ */
+
+package net.sourceforge.pmd.lang.apex.rule.bestpractices;
+
+import java.util.List;
+import net.sourceforge.pmd.lang.apex.ast.ASTMethod;
+import net.sourceforge.pmd.lang.apex.ast.ASTMethodCallExpression;
+import net.sourceforge.pmd.lang.apex.ast.ASTParameter;
+import net.sourceforge.pmd.lang.apex.ast.ASTUserClass;
+import net.sourceforge.pmd.lang.apex.rule.AbstractApexRule;
+
+/**
+ * Scans classes which implement the `Queueable` interface. If the `public void
+ * execute(QueueableContext context)` method does not call the `System.attachFinalizer(Finalizer f)`
+ * method, then a violation will be added to the `execute` method.
+ *
+ * @author mitchspano
+ */
+public class QueueableShouldAttachFinalizerRule extends AbstractApexRule {
+
+  private static final String EXECUTE = "execute";
+  private static final String QUEUEABLE = "queueable";
+  private static final String QUEUEABLE_CONTEXT = "queueablecontext";
+  private static final String SYSTEM_ATTACH_FINALIZER = "system.attachfinalizer";
+
+  /** Scans the top level class and all inner classes. */
+  @Override
+  public Object visit(ASTUserClass topLevelClass, Object data) {
+    scanClassForViolation(topLevelClass, data);
+    for (ASTUserClass innerClass : topLevelClass.descendants(ASTUserClass.class).toList()) {
+      scanClassForViolation(innerClass, data);
+    }
+    return data;
```

Comment:
This sounds like a [rule-chain](https://docs.pmd-code.org/latest/pmd_userdocs_extending_writing_java_rules.html#economic-traversal-the-rulechain) rule implementation. Therefore, it should register itself via 

```java
    @Override
    protected @NonNull RuleTargetSelector buildTargetSelector() {
        return RuleTargetSelector.forTypes(ASTUserClass.class);
    }
```

Then you can get rid of the manual for-loop (as the visit method will then be called for each ASTUserClass anyway).

Reply:
Great call, thanks for pointing this out! I've updated the implementation to use the rule chain.